torch/nn/functional.py

```diff
@@ -4916,11 +4916,11 @@ def _in_projection(
 
     # Efficient implementation equivalent to the following:
     scale_factor = 1 / math.sqrt(Q.size(-1)) if scale is None else scale
-    attn_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0) if is_causal else attn_mask
-    attn_mask = attn_mask.masked_fill(not attn_mask, -float('inf')) if attn_mask.dtype==torch.bool else attn_mask
+    attn_mask = torch.ones(L, S, dtype=torch.bool).tril(diagonal=0) if is_causal else torch.zeros(L, S, dtype=Q.dtype)
```

Comment:
is `Q.dtype` here safe? or should it be also torch.bool? (here's strange that `attn_mask.dtype` can either be bool or float

Reply:
if you look at the docs for the docs the mask can be either bool or of the same dtype as q,k,v. And if you look at the math/ref implementation you can see that boolean masks get converted to q.dtype during the execution of the function